### PR TITLE
feat: list posts, admin browse posts, user edit and delete post

### PR DESCRIPTION
**Pull request by gareth-go**
# Description
feature: show list post, admin browse post, edit post, delete post

# Screenshot
List post
image
Show post (if post is pending and user is admin, user can accept or refect it)
image
Show post (if current use is owner of the post it will have a button edit and delete, admin can also delete any post)
image
Edit post
image

# Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,7 +1,9 @@
 # frozen_string_literal: true
 
 Rails.application.routes.draw do
-  resources :posts, only: %i[new create show]
+  resources :posts
+  put 'posts/:id/accept', to: 'posts#accept', as: :accept_post
+  put 'posts/:id/reject', to: 'posts#reject', as: :reject_post
```
Comment (gary-gia-goldenowl) [inline config/routes.rb]: Use **member** method provided by Rails to wrap up the posts resources helps maintain a clear structure in your routes file

https://guides.rubyonrails.org/routing.html#adding-member-routes

File: app/views/posts/show.html.slim [hunk 2]
```diff
@@ -1,21 +1,46 @@
...
+
     .py-3.px-5
-      h4.fw-bold
-        = @post.user.username
+      span.fs-4.fw-bold
+        = render partial: "shared/post_username_link", locals: {post: @post}
+      -if @post.user == current_user || current_user.admin?
+        = link_to "Delete", @post, method: :delete, class: "btn btn-outline-secondary float-end"
+      -if @post.user == current_user
+        = link_to "Edit", edit_post_path(@post), class: "btn btn-outline-secondary float-end mx-2"
```
Comment (gary-gia-goldenowl) [inline app/views/posts/show.html.slim]: Fix space align

File: app/controllers/posts_controller.rb [hunk 3]
```diff
@@ -22,6 +36,42 @@ def create
     end
   end
 
+  def update
+    # remove blank tag value
+    # convert from tag_id to tag object
+    values = post_params
+    values[:tags].shift
+    values[:tags].map! { |tag| Tag.find(tag) }
```
Comment (gary-gia-goldenowl) [inline app/controllers/posts_controller.rb]: You can reuse these code line with private method, define a private method that will return params with tags something like (check feature logic carefully again)
```
private

def post_params_with_tags
    // Insert code here
end
```

File: app/controllers/posts_controller.rb [hunk 4]
```diff
@@ -1,19 +1,33 @@
...
+  def edit; end
+
   def create
     # remove blank tag value
     # convert from tag_id to tag object
     values = post_params
-    values['tags'].shift
-    values['tags'].map! { |tag| Tag.find(tag) }
+    values[:tags].shift
+    values[:tags].map! { |tag| Tag.find(tag) }
```
Comment (gary-gia-goldenowl) [inline app/controllers/posts_controller.rb]: You can reuse these code line with private method, define a private method that will return params with tags something like (check feature logic carefully again)
```
private

def post_params_with_tags
    // Insert code here
end
```

File: app/controllers/posts_controller.rb [hunk 5]
```diff
@@ -22,6 +36,42 @@ def create
...
+    @post.update(status: :rejected)
+
+    redirect_to @post
+  end
+
+  def destroy
+    @post.destroy
+
+    redirect_to posts_path
+  end
```
Comment (gary-gia-goldenowl) [inline app/controllers/posts_controller.rb:68]: Assume that all of these actions are just belongs to admin user, you can write a callback to secure the action for example:

```
before_action :require_admin, only: %i[accept reject destroy]


private

  def require_admin
    redirect_to root_path unless current_user.admin?
  end
```

File: app/controllers/posts_controller.rb [hunk 6]
```diff
@@ -1,19 +1,33 @@
 class PostsController < ApplicationController
   before_action :set_new_post, only: %i[create]
-  before_action :set_post, only: %i[show]
+  before_action :set_post, only: %i[show edit update accept reject destroy]
+
+  def index
+    # get all posts if current user is admin
+    # get only user's posts if current user is not admin
```
Comment (gary-gia-goldenowl) [inline app/controllers/posts_controller.rb]: Comment is very good practice but sometimes let the code speak for itself, you can read some best practices to write comment:

https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/

File: app/controllers/posts_controller.rb [hunk 7]
```diff
@@ -22,6 +36,42 @@ def create
...
+  def update
+    # remove blank tag value
+    # convert from tag_id to tag object
+    values = post_params
+    values[:tags].shift
+    values[:tags].map! { |tag| Tag.find(tag) }
+
+    # set cover_image nil if no image is attached
+    @post.cover_image.purge if values[:remove_cover_image] == '1' && !values.key?(:cover_image)
+    values.delete(:remove_cover_image)
```
Comment (nathan-phan-goldenowl) [inline app/controllers/posts_controller.rb]:
```
@post.cover_image.purge if params.dig(:posts, :remove_cover_image) == '1' && !values.key?(:cover_image)
```

File: app/controllers/posts_controller.rb [hunk 8]
```diff
@@ -1,19 +1,33 @@
...
+  def index
+    # get all posts if current user is admin
+    # get only user's posts if current user is not admin
+    @posts = if current_user.admin?
+               Post.all.order(created_at: :desc).includes(%i[user tags])
+             else
+               current_user.posts.order(created_at: :desc).includes(%i[tags])
+             end
+
+    @posts = @posts.where(status: Post.statuses[params[:status]]) if params[:status]
```
Comment (nathan-phan-goldenowl) [inline app/controllers/posts_controller.rb]:
```
@posts = @posts.where(status: Post.statuses[params[:status]]) if params[:status] && Post.statuses.include?(params[:status])
```

**Review by nathan-phan-goldenowl — APPROVED** [review]

**Pull request merged**